### ISSUE TITLE
Periodic pauses in cinnamon cause stutter in screen updates

### ISSUE THREAD

**Issue by sruggier** [linuxmint/cinnamon#8581]
* Cinnamon version: 4.0.10
 * Distribution: LMDE 3 (Cindy)
 * Graphics hardware: Nvidia Quadro K2100M w/ Nvidia driver
 * 64 bit

#### Issue

Cinnamon (the process/window manager) stutters periodically while updating the screen, about every 2 seconds on my system. My intuition was that a repeating pause like this is probably related to garbage collection, and some brief performance profiling seems confirm that (see the attachment below).

#### Steps to reproduce

* Log into a cinnamon session
* `apt-get install libmozjs-52-0-dbgsym libcjs-dbg linux-tools`
* Open a terminal and run `perf top -p $(pgrep -x cinnamon) -d1 -z` to see approximately what cinnamon is doing in any given second
* Move a window around, autoscroll, run glxgears. Anything that produces a smooth animation over time

#### Expected behaviour

Smooth animation, with no pauses or stuttering.

#### Actual behaviour

A pause that occurs about once every 2 seconds on my system. The `perf top` update immediately after the pause seems to show evidence of a GC operation taking place, with lots of time spent in symbols like `js::GCMarker::drainMarkStack`, `js::GCMarker::markAndScan<*>`, and so on. To illustrate what I mean, I've attached a snapshot of some perf terminal output at [cinnamon-perf-top-sample.txt](https://github.com/linuxmint/cinnamon/files/3213675/cinnamon-perf-top-sample.txt), from a one second window that included one of these pauses.

#### More info

I see this problem with browsers, 3D applications like glxgears, terminals, and the window manager itself, so I can rule out problems with any specific application. The perf output seems to confirm this.

**Comment by JosephMcc** [comment]
Do you have any sort of applet running that updates constantly like a system monitor applet?

**Comment by sruggier** [comment]
Yes, but I tried disabling those. The problem remained even after they were all removed from my panel. This isn't surprising, the file I attached basically rules out applets as the cause of the stutters: if something was blocking Cinnamon's UI thread without consuming CPU time, it wouldn't show up so prominently in `perf top`.

This seems to be vaguely driver related, the situation is much more tolerable when using the Intel driver (`Mesa DRI Intel(R) Haswell Mobile`, version 13.0.6). So I guess that somehow, while running on the Nvidia driver, Cinnamon is either using more GC-managed memory, or it's allocating more frequently.

**Comment by mcogoni** [comment]
Hi,
I'm using Mint 19.2 with kernel 4.15.0-70-generic on a Thinkpad T450S and an Intel GPU. I can reproduce the same behavior. The system works perfectly for hours or even days, then it starts to do the small pauses every few seconds for a fraction of a second.
The solution for me is to restart Cinnamon by doing Alt+F2 and executing the command "r".
Did anybody discover what's the root cause?

marco

**Comment by icarter09** [comment]
@mcogoni can you provide `/home/<username>/.xsession-errors` and  `/var/log/syslog`? Thanks.

**Comment by nalim** [comment]
I'have exactly the same problem (stuttering every 2s) and I tracked it down to a cinnamon desklet (not applet) called notes@schorschii (https://cinnamon-spices.linuxmint.com/desklets/view/38) which stupidly reloads and redraw all active notes (whenever changed or not)  EVERY 2s. Check you dekstop, maybe you the desklet there :-)

**Comment by SiegfredRodriguez** [comment]
Im currently having this problem in Arch, nvtop indicates sharp drops.

- It only happens when a second monitor is attached.

- Xorg logs indicated periodic disconnecting and reconnecting of displays causing xorg and or cinnamon to redraw.

These set of events happens every 10 seconds or so:

- Laptop monitor disconnecting 
- External Monitor disconnecting 
- Laptop Monitor reconnecting 
- External Monitor Reconnecting 

This happens quickly and looks like a stutter.

**Comment by macramole** [comment]
killing csd-power makes the stutter dissapear

activating `Perfom lid-closed action with external monitors attached` in the `Power Management setting` seems to be the workaround

https://unix.stackexchange.com/questions/739290/linux-mint-periodically-freezing-after-closing-lid

**Comment by Jazzhoo** [comment]
I've had that issue with Thinkpad T580. Laptop + external monitor (Lenovo usb-c dock gen 2). As that was >2 years ago I can't remember if I tried connecting the monitor on HDMI to rule out the dock issue. 

The cinnamon was nice and smooth after startup. However, as soon as I connected the dock and closed the lid, the system started to stutter. It was very annoying.. 

The ugly workaround was to have the lid slightly opened and manually set the external monitor as the only one.
I remember there was something about Thinkpads and low level lid closing event that didn't go well with Linux - I don't remember what that was.
